src/Microsoft.AspNetCore.Razor.Tools/ServerCommand.cs

```diff
@@ -89,5 +105,42 @@ protected virtual void ExecuteServerCore(ConnectionHost host, CompilerHost compi
             var dispatcher = RequestDispatcher.Create(host, compilerHost, cancellationToken, eventBus, keepAlive);
             dispatcher.Run();
         }
+
+        internal FileStream WritePidFile()
+        {
+            /*
+             * To make all the running rzc servers more discoverable, We want to write the process Id and pipe name to a file.
+             * The file contents will be in the following format,
+             *
+             * <PID>
+             * rzc
+             * <pipename>
+             */
+
+            const int DefaultBufferSize = 4096;
+            var processId = Process.GetCurrentProcess().Id;
+            var fileName = $"rzc-{processId}";
+
+            var path = Environment.GetEnvironmentVariable("DOTNET_BUILD_PIDFILE_DIRECTORY");
+            if (string.IsNullOrEmpty(path))
+            {
+                var homeEnvVariable = PlatformInformation.IsWindows ? "USERPROFILE" : "HOME";
+                path = Path.Combine(Environment.GetEnvironmentVariable(homeEnvVariable), ".dotnet/pids/build");
+            }
+
+            // Make sure the directory exists.
+            Directory.CreateDirectory(path);
+
+            path = Path.Combine(path, fileName);
+            var fileStream = new FileStream(path, FileMode.Create, FileAccess.Write, FileShare.Read, DefaultBufferSize, FileOptions.DeleteOnClose);
```

Comment:
Is delete on close trustworthy? I thought there was some doubt about this.

Reply:
From my local testing this works consistently. This was also mentioned in the design doc. @peterhuene thoughts?

Reply:
Ultimately I'd have to take a look at the framework, but I would wager that this is cross-plat without issue.  On Windows, I assume it uses `FILE_FLAG_DELETE_ON_CLOSE`. On macOS and Linux, I assume the file is immediately unlinked.

Reply:
Unfortunately I was wrong about this, at least on macOS and Linux: there is no analogous mechanism to `FILE_FLAG_DELETE_ON_CLOSE` for those platforms (and of course immediate unlinking would only work if you don't expect another process to be able to read the file).  While this works for Windows as intended, sending a `SIGINT` or `SIGTERM` to the server process on the other platforms leaves the pid file behind.

There are a few ways to handle these signals in a cross-platform way with corefx:
Handle SIGTERM with `AssemblyLoadContext.Default.Unloading`
Handle SIGINT/SIGQUIT with `Console.CancelKeyPress`

I can handle cleaning up defunct pid files from the `build-server shutdown` command, but since this is a non-tmpfs location, we should probably also fix it in the server.

Obviously this won't fix SIGKILL, but caveat emptor when sending that signal.

My apologies for not realizing this sooner.